src/lib/comms/sol-http-client-impl-curl.c

```diff
@@ -0,0 +1,840 @@
+        return;
+
+    if (connection->error) {
+        param = NULL;
+        goto out;
+    }
+
+    param = &(struct sol_http_response) {
+        .api_version = SOL_HTTP_RESPONSE_API_VERSION,
+        .content = connection->buffer
+    };
+
+    r = curl_easy_getinfo(connection->curl, CURLINFO_CONTENT_TYPE, &tmp);
+    if (r == CURLE_OK) {
+        param->content_type = tmp ? strdupa(tmp) : "application/octet-stream";
+        r = curl_easy_getinfo(connection->curl, CURLINFO_EFFECTIVE_URL, &tmp);
+    }
+    if (r == CURLE_OK && tmp) {
+        long response_code;
+
+        param->url = strdupa(tmp);
+        r = curl_easy_getinfo(connection->curl, CURLINFO_RESPONSE_CODE,
+            &response_code);
+        if (r == CURLE_OK)
+            param->response_code = (int)response_code;
+    }
+    if (r != CURLE_OK)
+        param = NULL;
+
+out:
+    connection->cb((void *)connection->data, param);
+    connection->cb = NULL; /* Don't call again. */
+}
+
+static size_t
+write_cb(char *data, size_t size, size_t nmemb, void *connp)
+{
+    struct connection *connection = connp;
+    size_t data_size;
+    int r;
+
+    r = sol_util_size_mul(size, nmemb, &data_size);
+    SOL_INT_CHECK(r, < 0, 0);
+
+    r = sol_buffer_append_slice(&connection->buffer,
+        SOL_STR_SLICE_STR(data, data_size));
+    SOL_INT_CHECK(r, < 0, 0);
+
+    return data_size;
+}
+
+static void
+pump_multi_info_queue(void)
+{
+    CURLMsg *msg;
+    int msgs_left;
+
+    while ((msg = curl_multi_info_read(global.multi, &msgs_left))) {
+        struct connection *conn;
+
+        if (msg->msg != CURLMSG_DONE)
+            continue;
+
+        curl_easy_getinfo(msg->easy_handle, CURLINFO_PRIVATE, &conn);
+        call_connection_finish_cb(conn);
+    }
+}
+
+static bool
+multi_perform_cb(void *data)
+{
+    int running;
+
+    if (!global.multi)
+        goto out;
+
+    pump_multi_info_queue();
+
+    if (curl_multi_perform(global.multi, &running) == CURLM_OK) {
+        if (running > 0)
+            return true;
+    }
+
+out:
+    global.multi_perform_timeout = NULL;
+    return false;
+}
+
+static int
+timer_cb(CURLM *multi, long timeout_ms, void *userp)
+{
+    if (timeout_ms > 0) {
+        if (timeout_ms < 100)
+            timeout_ms = 100;
+        if (global.timeout_ms == timeout_ms)
+            return 0;
+
+        /* cURL requested a timeout value change. */
+        global.timeout_ms = timeout_ms;
+
+        if (global.multi_perform_timeout) {
+            /* Change sol_timeout if there's already one in place. */
+            sol_timeout_del(global.multi_perform_timeout);
+            global.multi_perform_timeout = sol_timeout_add(global.timeout_ms,
+                multi_perform_cb, NULL);
+
+            return global.multi_perform_timeout ? 0 : -1;
+        }
+    } else if (!timeout_ms) {
+        /* Timer expired; pump cURL immediately. */
+        return multi_perform_cb(NULL) ? 0 : -1;
+    }
+
+    return 0;
+}
+
+static bool
+http_init(void)
```

Comment:
This function will probably be called `sol_comms_init()` (to be called by `sol_init()`). Shutting down cURL every time is not a good idea as it caches DNS queries and other things. Same for `http_shutdown()`. I'll make another PR for this if this patch series is OK; otherwise, I'll submit the changes in the next revision of this PR.